litmus-concepts.md

```diff
@@ -0,0 +1,311 @@
+  * Example: custom SSH key/options
+  * Bolt has now plugins to receive inventory information from outside sources - would this be a good way to keep dynamic litmus data out of inventory.yml?
+
+* Find ways to pass arguments to provisioners
+  * Hypervisor Credentials and arguments
+  * bolt options (e.g custom SSH key/options)
+  * which parts of this are module/repo specific? which parts need to be per-user?
+
+
+## Test Infrastructure
+
+For full system-level testing of acceptance criteria, tests require access to systems to test.
+Depending on the use-case and resources available, this test infrastructure can be accessed in a variety of ways.
+There are a few necessary conditions for all those test systems:
+* accessible through bolt - this allows tests to act on such a system
+* initially clean - without a well-defined initial state, tests become complex, unreliable, or both
+* dedicated - running tests on shared systems is inviting troubles, either by tests stepping on each other or the tests interfering with user activities or the other way around
+* representative of production systems - running the tests should provide insight about expected behaviour in the real world
+
+### What can be used as Test Infrastructure?
+
+Throw-away VMs are the easiest way to fulfil those conditions.
+Provisioned from upstream images or organisations' golden templates,
+they provide a complete operating system, accurate representation of production and full isolation.
+The downside of virtual machines are the high resource usage and provisioning times in the order of minutes.
+You can use local VMs on your development workstation, or private and public cloud providers.
+
+To reduce resource usage and provisioning times docker and containers come in handy.
+They deploy in seconds and achieve high packing densities with low overhead.
+Best practices for container images justifiably frowns on SSH access or complete operating system services,
+thus common public images are usually not representative of production on full VMs.
+To avoid this limitation, [puppetlabs/litmusimages](https://github.com/puppetlabs/litmusimage) provides a set of pre-fabbed [docker images](https://hub.docker.com/u/litmusimage) that allow SSH access and have an init process running.
+
+In some cases, using bare metal servers or already running systems is unavoidable.
+
+### How to provision Test Infrastructure
+
+There are as many ways to aqcuire access to test systems as there are kinds of test systems.
+By default, Litmus calls out to a provisioner task to provision or tear down systems on demand.
+In the `puppetlabs-provision` module, we ship a number of commonly needed provisioners for docker (with SSH), docker_exp (using bolt's docker transport), vagrant (for local VMs) and the private cloud APIs [VMpooler](https://github.com/puppetlabs/vmpooler) and [ABS](https://github.com/puppetlabs/always-be-scheduling).
```

Comment:
ABS is a private repository with a private tool designed for this provisioning, not sure if it's necessary to mention it

Reply:
```suggestion
In the `puppetlabs-provision` module, we ship a number of commonly needed provisioners for docker (with SSH), docker_exp (using bolt's docker transport), vagrant (for local VMs) and the private cloud APIs [VMpooler](https://github.com/puppetlabs/vmpooler) and the puppet-private [ABS](https://github.com/puppetlabs/always-be-scheduling).
```

Reply:
I'll leave it in but mention that it is puppet-private.